### DISCUSSION TITLE
Performance depending on the number and order of filters?

### DISCUSSION THREAD

**Discussion by jessicamrbr** [tekartik/sembast.dart#370] [Q&A] [answered]
Is there any performance gain in the order of applying filters? Is there any negative or positive impact depending on the number of filters used?

**Comment by alextekartik** [comment, marked as answer]
There is no indexing so indeed the more filters you have the slower it gets as every record in a given store will be checked. The algorithm is dumb so indeed the order could have some impact (in a AND the first filter will be checked first). If you really want to optimize the filter, the faster will be to use a [custom filter](https://pub.dev/documentation/sembast/latest/sembast/Filter/Filter.custom.html) where you access the raw data of each record directly.

A query with complex filtering with lots of data could be slow for each [getSnapshots](https://pub.dev/documentation/sembast/latest/sembast/SembastQueryRefExtension/getSnapshots.html) call, however [onSnapshots](https://pub.dev/documentation/sembast/latest/sembast/SembastQueryRefExtension/onSnapshots.html) will be smart enough to only check the filtering on modified records.

**Reply by jessicamrbr** [reply]
Thank you very much for the extremely quick response!

Sorry to ask without analyzing first. I noticed that there were no indexes other than the store name + key. But I didn't stop to look in depth at how the filters work.

But from your answer I understand that it goes through all the documents in a single Loop, and applies the conditions in the order of registration. I imagine that if it is an AND and the first condition is not met, it does not validate the others, correct?

The doubt arose when I was combining a CUSTON filter with other filters. Whether it would be more advantageous to write other conditions within the custon or use the methods already available for pre-filtering. I don't think it will make a difference, at most I pay attention to the use of AND /OR and their order.

**Comment by alextekartik** [comment]
> I imagine that if it is an AND and the first condition is not met, it does not validate the others, correct?

That's correct. And in the same way if the first condition of a OR is met it does not check the others too.
